### PR TITLE
Replace trending content with most popular

### PR DESCRIPTION
**Pull request by Rosa-Fox**
[Trello](https://trello.com/c/SalBDmy9/242-display-screen-work-reinstate-middle-column)

The middle column of the GOV.UK display screen was blank. This was because [the page view data was previously accessed from an API provided by the Performance Platform](https://github.com/alphagov/govuk-display-screen/blob/main/public/javascripts/content.js#L12) which has been retired. 

In this PR, the Google Analytics Core Reporting API is queried from within this application and the formatted response is exposed as a JSON endpoint. 

Previously the middle column displayed trending GOV.UK content. This has been changed to instead display the most popular content. We felt this would be more appropriate as the intention is that this dashboard is displayed around the office, so 'most popular' is much easier to interpret at a quick glance. The query for working out the trends was much more expensive and involved a lot of data processing, which seemed excessive for this use case.

Please note that no styling has been added, it just displays the data. In a following PR, I will add the design system into this application and improve the styling.

The longer term future of the display screen is being discussed and it is quite possible that it will be rebuilt from scratch. This is likely to be a stop gap whilst we build something better!

To the reviewer: you will need to create a `.env` file and add the credentials. I'll add them to the live server so that you can access them.

Screenshot 2022-10-31 at 14 31 23

Screenshot 2022-10-31 at 14 31 36

### CODE REVIEW

File: content.rb [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  attr_reader :service, :credentials
+  def initialize
+    @service = Google::Apis::AnalyticsreportingV4::AnalyticsReportingService.new
+    @credentials = Google::Auth::ServiceAccountCredentials.make_creds(
+      scope: "https://www.googleapis.com/auth/analytics.readonly"
+    )
+    @service.authorization = @credentials
+  end
+
+  # Return formatted page data for 10 most popular pages on GOV.UK
```
Comment (gclssvglx) [inline content.rb:15]: I'm not sure we need this comment - the method name says it all for me.

File: content.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  def most_popular_govuk_pages
+    format_response.to_json
+  end
+
+private
+
+  # Make request to Google Analytics Core Reporting API 
+  def ga_request
```
Comment (gclssvglx) [inline content.rb:23]: Maybe `google_analytics_request` 🤔 then we can remove the comment maybe

File: content.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  end
+
+  # Build Google Analytics report request
+  def analytics_reports
+    Google::Apis::AnalyticsreportingV4::GetReportsRequest.new(
+      { report_requests: [build] }
+    )
+  end
+
+  # Format response
```
Comment (gclssvglx) [inline content.rb:34]: I don't think we need this comment.

File: content.rb [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            page_title: row[:dimensions][1]
+          }
+          page_data << row_data
+        end
+      page_data
+    end
+  end
+
+  # Build the GA query
+  def build
```
Comment (gclssvglx) [inline content.rb:62]: Maybe `build_google_analytics_query` and remove the comment 🤔

File: content.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      date_ranges: [date_range],
+      metrics: [metric],
+      order_bys: [{field_name: 'ga:pageviews', sort_order: 'DESCENDING'}],
+      dimensions: [dimension_path, dimension_title],
+      page_size: '10'
+    )
+  end
+
+  # Set the date range
+  def date_range
```
Comment (gclssvglx) [inline content.rb:75]: `set_date_range` maybe 🤔

File: content.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    start_date = Date.today - 6
+    end_date = Date.today
+
+    Google::Apis::AnalyticsreportingV4::DateRange.new(
+      start_date: start_date.to_s,
+      end_date: end_date.to_s
+    )
+  end
+
+  # Set the metric
```
Comment (gclssvglx) [inline content.rb:85]: Similar suggestions to the other methods - could we rename and remove the need for the comments?

File: content.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    )
+  end
+
+  # Set the pageTitle dimension
+  def dimension_title
+    Google::Apis::AnalyticsreportingV4::Dimension.new(
+      name: "ga:pageTitle"
+    )
+  end
+end
```
Comment (gclssvglx) [inline content.rb:105]: Looks like we need an empty last line

File: server.rb [hunk 8]
```diff
@@ -54,6 +55,13 @@
   JSON.generate Hash.from_xml(response.body)
 end
 
+get '/popular-content' do
+  content_type :json
+
+  content = Content.new()
+  content.most_popular_govuk_pages
+end
+
```
Comment (gclssvglx) [inline server.rb:64]: Nice! I love this use of an internal route - that calls an internal class - as a substitute for calling an external API directly. Very clean 👏

File: public/index.html [hunk 9]
```diff
@@ -94,4 +95,4 @@ <h1>Recent content from departments:</h1>
       </div>
     </li>
   {{/tweets}}
-</script>
+</script>
```
Comment (gclssvglx) [inline public/index.html:98]: Think we're missing an empty last line

File: public/javascripts/content.js [hunk 10]
```diff
@@ -47,4 +38,4 @@
   };
 
   root.matrix.content = content;
-}).call(this);
+}).call(this);
```
Comment (gclssvglx) [inline public/javascripts/content.js:41]: Think we're missing an empty last line

File: public/javascripts/content.js [hunk 11]
```diff
@@ -9,25 +9,16 @@
...
-          title = data.data[i].pageTitle.split(' | ').slice(0,-1).join(' - ');
-        }
-        content.pages.push({
-          title: title,
-          displayHits: root.matrix.numberWithCommas(data.data[i].week2),
-          percentageUp: root.matrix.numberWithCommas(Math.round(data.data[i].percent_change)) + "%"
+      for (let i = 0; i < data.length; i++) {
+        content.pages.push ({
+          title: data[i]["page_title"],
+          displayHits: data[i]["page_views"],
```
Comment (gclssvglx) [inline public/javascripts/content.js:19]: So much nicer - thank you

**Review by gclssvglx — COMMENTED** [review]
Great work @Rosa-Fox 🥇  I'm being super picky about some comments and missing empty lines - sorry. I really like the use of an internal route to encapsulate an external API call very nice. Might nick that idea 😁

**Comment by Rosa-Fox** [comment]
@gclssvglx thanks! I made the changes so its ready for re-review.

**Pull request merged**